runtime/src/accounts.rs

```diff
@@ -739,65 +659,66 @@ impl Accounts {
+        let accounts_db = Arc::new(AccountsDB::new(&paths));
         Accounts {
             accounts_db,
-            account_locks: Mutex::new(HashMap::new()),
+            account_locks: Mutex::new(HashSet::new()),
             paths,
             own_paths,
         }
     }
-
-    pub fn new_from_parent(&self, fork: Fork, parent: Fork) {
-        self.accounts_db.add_fork(fork, Some(parent));
-    }
-
-    /// Slow because lock is held for 1 operation instead of many
-    pub fn load_slow(&self, fork: Fork, pubkey: &Pubkey) -> Option<Account> {
-        self.accounts_db
-            .load(fork, pubkey, true)
-            .filter(|acc| acc.lamports != 0)
+    pub fn new_from_parent(parent: &Accounts) -> Self {
+        let accounts_db = parent.accounts_db.clone();
+        Accounts {
+            accounts_db,
+            account_locks: Mutex::new(HashSet::new()),
+            paths: parent.paths.clone(),
+            own_paths: parent.own_paths,
+        }
     }
 
     /// Slow because lock is held for 1 operation instead of many
-    pub fn load_slow_no_parent(&self, fork: Fork, pubkey: &Pubkey) -> Option<Account> {
+    pub fn load_slow(&self, ancestors: &HashMap<Fork, usize>, pubkey: &Pubkey) -> Option<Account> {
         self.accounts_db
-            .load(fork, pubkey, false)
+            .load_slow(ancestors, pubkey)
             .filter(|acc| acc.lamports != 0)
     }
 
-    /// Slow because lock is held for 1 operation instead of many
-    pub fn load_by_program_slow_no_parent(
-        &self,
-        fork: Fork,
-        program_id: &Pubkey,
-    ) -> Vec<(Pubkey, Account)> {
-        self.accounts_db
-            .load_by_program(fork, program_id, false)
+    pub fn load_by_program(&self, fork: Fork, program_id: &Pubkey) -> Vec<(Pubkey, Account)> {
+        let accumulator: Vec<Vec<StoredAccount>> = self.accounts_db.scan_account_storage(
+            fork,
+            |stored_account: &StoredAccount, accum: &mut Vec<StoredAccount>| {
+                if stored_account.account.owner == *program_id {
+                    accum.push(stored_account.clone())
+                }
+            },
+        );
+        let mut versions: Vec<StoredAccount> = accumulator.into_iter().flat_map(|x| x).collect();
+        versions.sort_by_key(|s| (s.pubkey, (s.write_version as i64).neg()));
+        versions.dedup_by_key(|s| s.pubkey);
+        versions
             .into_iter()
-            .filter(|(_, acc)| acc.lamports != 0)
+            .map(|s| (s.pubkey, s.account))
             .collect()
     }
 
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
-        self.accounts_db.store(fork, pubkey, account);
+        self.accounts_db.store(fork, &[(pubkey, account)]);
     }
 
     fn lock_account(
-        fork: Fork,
-        account_locks: &mut HashMap<Fork, HashSet<Pubkey>>,
+        locks: &mut HashSet<Pubkey>,
```

Comment:
how does this work if replay stage and banking_stage are going after the same account on different slots?

Reply:
nevermind, bank.accounts is a new struct every time